sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala

```diff
@@ -1091,13 +1102,49 @@ class SessionCatalog(
   // ----------------------------------------------------------------
 
   /**
+   * Constructs a [[FunctionBuilder]] based on the provided class that represents a function.
+   */
+  protected def makeFunctionBuilder(name: String, functionClassName: String): FunctionBuilder = {
+    val clazz = Utils.classForName(functionClassName)
+    (children: Seq[Expression]) => {
+      try {
+        makeFunctionExpression(name, Utils.classForName(functionClassName), children).getOrElse {
+          val extraMsg =
+            if (!isUsingHiveMetastore) "Use sparkSession.udf.register(...) instead." else ""
+          throw new AnalysisException(
+            s"No handler for UDF/UDAF/UDTF '${clazz.getCanonicalName}'. $extraMsg")
+        }
+      } catch {
+        case ae: AnalysisException =>
+          throw ae
+        case NonFatal(e) =>
+          val analysisException =
+            new AnalysisException(s"No handler for UDF/UDAF/UDTF '${clazz.getCanonicalName}': $e")
+          analysisException.setStackTrace(e.getStackTrace)
+          throw analysisException
+      }
+    }
+  }
+
+  /**
    * Construct a [[FunctionBuilder]] based on the provided class that represents a function.
    *
    * This performs reflection to decide what type of [[Expression]] to return in the builder.
    */
-  protected def makeFunctionBuilder(name: String, functionClassName: String): FunctionBuilder = {
-    // TODO: at least support UDAFs here
-    throw new UnsupportedOperationException("Use sqlContext.udf.register(...) instead.")
+  protected def makeFunctionExpression(
```

Comment:
seems we need to catch exception for this method anyway, how about we just make this method return `Expression` and document that it can throw exception if the given class is not supported? Then `HiveSessionCatalog` can define its own exception message.